src/Ztm.WebApi/TransactionConfirmationWatcher.cs

```diff
@@ -0,0 +1,354 @@
+                timer.Elapsed += OnTimeout;
+                var due = watch.Due - DateTime.UtcNow;
+                timer.Start(due < TimeSpan.Zero ? TimeSpan.Zero : due, null, watch.Id);
+            }
+            finally
+            {
+                timerLock.ExitWriteLock();
+            }
+        }
+
+        async void OnTimeout(object sender, TimerElapsedEventArgs e)
+        {
+            var watch = await watchRepository.GetAsync((Guid)e.Context, CancellationToken.None);
+
+            await Execute(watch.Callback, watch.Timeout, CancellationToken.None);
+            RemoveTimer(watch.Transaction, watch.Id);
+        }
+
+        void RemoveTimer(uint256 transaction, Guid id)
+        {
+            timerLock.EnterWriteLock();
+
+            try
+            {
+                timers[transaction].Remove(id);
+                if (timers[transaction].Count == 0)
+                {
+                    timers.Remove(transaction);
+                }
+            }
+            finally
+            {
+                timerLock.ExitWriteLock();
+            }
+        }
+
+        async Task<bool> StopTimer(uint256 transaction, Guid id)
+        {
+            timerLock.EnterWriteLock();
+            try
+            {
+                if (timers.TryGetValue(transaction, out var txTimers) && txTimers.TryGetValue(id, out var timer))
+                {
+                    await timer.Item1.StopAsync(CancellationToken.None);
+                    if (timer.Item1.ElapsedCount == 0)
+                    {
+                        RemoveTimer(transaction, id);
+                        return true;
+                    }
+                }
+            }
+            finally
+            {
+                timerLock.ExitWriteLock();
+            }
+
+            return false;
+        }
+
+        void ResumeTimer(ConfirmContext watch)
+        {
+            timerLock.EnterWriteLock();
+
+            try
+            {
+                TimerSetup(watch);
+            }
+            finally
+            {
+                timerLock.EnterWriteLock();
+            }
+        }
+
+        async Task Confirm(TransactionWatch<ConfirmContext> watch)
+        {
+            await Execute(watch.Context.Callback, watch.Context.Success, CancellationToken.None);
+        }
+
+        async Task Execute(Callback callback, TransactionConfirmationCallbackResult payload, CancellationToken cancellationToken)
+        {
+            await callbackRepository.AddHistoryAsync(callback.Id, payload, cancellationToken);
+
+            if (await this.callbackExecuter.Execute(callback.Id, callback.Url, payload, cancellationToken))
+            {
+                await callbackRepository.SetCompletedAsyc(callback.Id, cancellationToken);
+            }
+        }
+
+        // IBlockListener
+
+        public Task BlockAddedAsync(Block block, int height, CancellationToken cancellationToken)
```

Comment:
The implementation of `IBlockListener` should be explicitly due to the user of `TransactionConfirmationWatcher` don't need to call into those methods.